servant/src/Servant/API/Authentication.hs

```diff
@@ -0,0 +1,41 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE KindSignatures     #-}
+{-# LANGUAGE PolyKinds          #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# OPTIONS_HADDOCK not-home    #-}
+module Servant.API.Authentication
+( AuthPolicy (..)
+, AuthProtect (..)
+, AuthProtected (..)
+, BasicAuth (..)
+, JWTAuth (..)
+) where
+
+
+import           Data.ByteString (ByteString)
+import           Data.Typeable   (Typeable)
+import           GHC.TypeLits    (Symbol)
+import           Data.Text       (Text)
+
+-- | we can be either Strict or Lax.
+-- Strict: all handlers under 'AuthProtect' take a 'usr' argument.
+--         when auth fails, we call user-supplied handlers to respond.
+-- Lax: all handlers under 'AuthProtect' take a 'Maybe usr' argument.
+--      when auth fails, we call the handlers with 'Nothing'.
+data AuthPolicy = Strict | Lax
+
+-- | the combinator to be used in API types
+data AuthProtect authdata usr (policy :: AuthPolicy)
+
+-- | what we'll ask user to provide at the server-level when we see a
+-- 'AuthProtect' combinator in an API type
+data family AuthProtected authdata usr subserver :: AuthPolicy -> *
```

Comment:
E.g.:
```
data AuthStrictness = StrictAuth | LaxAuth
data OnMissing  a where
    LaxMissing :: OnMissing Lax
    StrictMissing :: IO ServantErr -> OnMissing Strict

data OnUnauthenticated authData stricness where
   LaxUnauthenticated :: OnUnathenticated Lax
   StrictUnauthenticated :: (authData -> IO ServantErr) -> OnUnauthenticated Strict


data AuthProtected authData missingStrictness unauthStrictness subserver = AuthProtected
  { onMissing :: OnMissing missingStrictness
  , onUnathenticated :: OnUnathenticated authData unauthStrictness
  , subserver :: subserver -- but note commment below
  }

```

Re: `subserver`. The analogue of `->` is `(,)`, so I wonder if we're doing the right thing by having the subserver insider the `AuthProtected` datatype. If we want to keep the server regular, I think `(,)` makes more sense. This comes up right now already with e.g. `enter`. @alpmestan what do you think?